src/utils/mtev_lru.c

```diff
@@ -0,0 +1,252 @@
+#include "mtev_lru.h"
+#include "mtev_hash.h"
+#include "mtev_log.h"
+
+#include <ck_hs.h>
+#include <sys/queue.h>
+
+struct lru_entry {
+  void *entry;
+  size_t key_len;
+  TAILQ_ENTRY(lru_entry) list_entry;
+  char key[];
+};
+
+struct mtev_lru {
+  ck_hs_t hash;
+  int32_t max_entries;
+  TAILQ_HEAD(lru_list, lru_entry) lru_cache;
+  int32_t lru_cache_size;
+  uint32_t expire_count;
+  void (*free_fn)(void *);
+  pthread_mutex_t mutex;
+};
+
+static void *
+lru_malloc(size_t r)
+{
+  return malloc(r);
+}
+
+static void
+lru_free(void *p, size_t b, bool r)
+{
+  (void)b;
+  (void)r;
+  free(p);
+  return;
+}
+
+static struct ck_malloc malloc_ck_hs = {
+  .malloc = lru_malloc,
+  .free = lru_free
+};
+
+static int rand_init = 0;
+
+#define HS_HF_STRING(F)                         \
+  static unsigned long                          \
+  F(const void *k, unsigned long seed)          \
+  {                                             \
+    return mtev_hash__hash(k, strlen(k), seed);   \
+  }
+
+HS_HF_STRING(lru_entry_hash);
+
+#define offset_of(type, member)                 \
+  ((size_t)(&((type *)0)->member))
+
+#define container_of(derived_ptr, type, field)                \
+  ((type *)((char *)(derived_ptr) - offset_of(type, field)))
+
+static void
+hs_init(ck_hs_t *hs, unsigned int mode, ck_hs_hash_cb_t *hf, ck_hs_compare_cb_t *cf, unsigned long size)
+{
+  if(!rand_init) {
+    srand48((long int)time(NULL));
+    rand_init = 1;
+  }
+  struct ck_malloc *allocator = &malloc_ck_hs;
+
+  if (ck_hs_init(hs, mode | CK_HS_MODE_SPMC, hf, cf, allocator, size, lrand48()) == false) {
+    mtevFatal(mtev_error, "Cannot initialize ck_hs\n");
+  }
+}
+
+static inline bool
+hs_string_compare(const void *a, const void *b)
+{
+  return strcmp((const char * const)a, (const char * const)b) == 0;
+}
+
+mtev_lru_t *
+mtev_lru_create(int32_t max_entries, void (*free_fn)(void *)) 
+{
+  struct mtev_lru *r = malloc(sizeof(struct mtev_lru));
+  if (max_entries <= 0) {
+    hs_init(&r->hash, CK_HS_MODE_OBJECT | CK_HS_MODE_DELETE, lru_entry_hash, hs_string_compare, 1024);
+  } else {
+    hs_init(&r->hash, CK_HS_MODE_OBJECT | CK_HS_MODE_DELETE, lru_entry_hash, hs_string_compare, max_entries * 2);
+  }
+  TAILQ_INIT(&r->lru_cache);
+  r->lru_cache_size = 0;
+  r->expire_count = 0;
+  r->max_entries = max_entries;
+  if (free_fn != NULL) {
+    r->free_fn = free_fn;
+  } else {
+    r->free_fn = free;
+  }
+  pthread_mutex_init(&r->mutex, NULL);
+  return r;
+}
+
+void
+mtev_lru_destroy(mtev_lru_t *lru) 
+{
+  mtev_lru_invalidate(lru);
+  ck_hs_destroy(&lru->hash);
+  free(lru);
+}
+
+void
+mtev_lru_invalidate(mtev_lru_t *lru)
+{
+  pthread_mutex_lock(&lru->mutex);
+  void *entry;
+  ck_hs_iterator_t it = CK_HS_ITERATOR_INITIALIZER;
+  while(ck_hs_next(&lru->hash, &it, &entry) == true) {
+    struct lru_entry *e = container_of(entry, struct lru_entry, key);
+    lru->free_fn(e->entry);
+    free(e);
```

Comment:
looks like this will `free` the entry for the `ck_hs`, but we haven't actually updated the `ck_hs` to remove the entry? this is fine when `mtev_lru_invalidate` is called by `mtev_lru_destroy`, but not fine when `mtev_lru_invalidate` is called otherwise...